engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,349 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.ibm.icu.text.Normalizer2;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.api.profiles.LoopConditionProfile;
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+
+/**
+ * Implements {@code hash_code} functionality.
+ *
+ * <p>Make sure that the hashing contract is retained after any modification.
+ *
+ * <h3>Hashing contract:</h3>
+ *
+ * <ul>
+ *   <li>Whenever two objects are equal ({@code EqualsAnyNode} returns {@code true}), their hashcode
+ *       should equal. More formally: {@code For all objects o1, o2: if o1 == o2 then hash(o1) ==
+ *       hash(o2)}
+ *   <li>Whenever two hash codes are different, their associated objects are different: {@code For all objects
+ *       o1, o2: if hash(o1) != hash(o2) then o1 != o2.
+ * </ul>
+ */
+@BuiltinMethod(type = "Any", name = "hash_code", description = "Returns hash code of the object.")
+@GenerateUncached
+public abstract class HashCodeAnyNode extends Node {
+
+  public static HashCodeAnyNode build() {
+    return HashCodeAnyNodeGen.create();
+  }
+
+  public abstract long execute(@AcceptsError Object self);
+
+  /** Specializations for primitive values * */
+  @Specialization
+  long hashCodeForShort(short s) {
+    return s;
+  }
+
+  @Specialization
+  long hashCodeForByte(byte b) {
+    return b;
+  }
+
+  @Specialization
+  long hashCodeForLong(long l) {
+    return l;
+  }
```

Comment:
```suggestion
  long hashCodeForLong(long l) {
    return Long.hashCode(l);
  }
```

Why not use that? As you mentioned, the actual hashcode is an `int` so half bits of that `long` will be truncated later, right? If we return the `long` as-is, half of its information will be discarded for the hashcode.